core/pollfds: Use a compact pollfd array

shefty opened the pull request:
The mdtest benchmark is showing significant performance regressions
    as a result of commit 7ec82992.  That commit replaced a compact
    pollfd array, where all entries are valid, with a sparsely filled
    array.  This removed a linear search through the array in order
    to update events.
    
    Rather than revert the above commit, we separate the pollfd
    context array and pollfd array.  The fd may still be used to
    directly index into the context array.  The context structure is
    updated to store the corresponding index where the fd is located
    in the pollfd array.  This avoids the linear search while
    maintaining a compact array.
    
    As a result of these changes, it becomes possible to support
    the hot fd array on windows.  Add that as part of this change in
    order to align the unix and windows implementations closer.
    
    Signed-off-by: Sean Hefty <sean.hefty@intel.com>

shefty commented:
This series results in frequent hangs in the socket provider running to cq_data test, which is one of the first tests that uses blocking cq read calls.  Debugging.

j-xiong commented on src/unix/osd.c:
```diff
@@ -316,51 +316,72 @@ int ofi_set_thread_affinity(const char *s)
-		pfds->nfds--;
+	pfds->nfds--;
+	if (ctx->index < pfds->nfds) {
+		pfds->fds[ctx->index] = pfds->fds[pfds->nfds];
+
+		swap_ctx = ofi_pollfds_get_ctx(pfds, pfds->fds[pfds->nfds].fd);
+		swap_ctx->index = ctx->index;
```
Simply swapping L377 and L379 would cause `ofi_pollfds_get_ctx` to return NULL since `pfds->nfds` is outside the loop boundary.

shefty replied:
gah - yep, no wonder windows is still broken.

shefty commented:
Updated --  merged the windows and unix implementations around pollfds add/modify/del into the common code.  Added a new OS specific ofi_pollfds_alloc_ctx() helper in their place, and updated the ofi_pollfds_get_ctx() to handle error checking on the fd.  Because of these changes I was able to add a few more asserts to the code.  Fixed the ofi_pollfds_do_del() path

shefty commented on src/common.c:
```diff
@@ -1696,16 +1747,19 @@ static int ofi_pollfds_waitall(struct ofi_pollfds *pfds,
 				events[found].events = pfds->fds[i].revents;
-				events[found++].data.ptr = pfds->ctx[i].context;
+				ctx = ofi_pollfds_get_ctx(pfds, pfds->fds[i].fd);
+				assert(ctx && ctx->context);
```
I thought this would be a nice check, but it turns out that the socket provider sets the context to NULL in some cases.  The sockets provider uses epoll where available, but pollfds where not.  This can be hit by using configure --disable-epoll to force this path.

shefty commented:
Intel CI aborted for oneCCL with rxm;tcp, but I didn't see any actual log output.  Appveyor failure:
```
bw -e msg -v -T 1
timeout, killing job
SERVER OUTPUT: 
----
CLIENT OUTPUT: 
----
test failed due to timeout
scalable_ep
SERVER OUTPUT: 
wait for recv completion for ctx: 0
failed compare expected 0x12345670, read 0x12345671
wait for recv completion for ctx: 1
----
CLIENT OUTPUT: 
Posting send for ctx: 0
Posting send for ctx: 1
----
==============================
short tests:
msg_pingpong -I 5
SERVER OUTPUT: 
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      5       640         0.14s      0.00   14062.80       0.00
256     5       2.5k        0.10s      0.03    9999.90       0.00
1k      5       10k         0.32s      0.03   3[2313](https://ci.appveyor.com/project/ofiwg/libfabric/builds/43232829/job/f71pk7vllanp6qgg#L2313).40       0.00
4k      5       40k         0.32s      0.13   31967.40       0.00
64k     5       640k        0.28s      2.32   28199.90       0.00
1m      5       10m         0.28s     37.35   28074.00       0.00
----
CLIENT OUTPUT:
```
The bw test failed.  But looking at the results for other tests, the performance is really bad.

shefty commented:
Swapping in the windows implementation in place of the unix one, I can hit this assert for fi_bw test:
```
fi_bw: src/common.c:1751: ofi_pollfds_waitall: Assertion `ctx' failed.

#4  0x00007ffff7908286 in ofi_pollfds_waitall (pfds=0x5555557d1910, 
    events=0x5555557cd900, maxevents=2, timeout=0) at src/common.c:1751
#5  0x00007ffff790847b in ofi_pollfds_wait (pfds=0x5555557d1910, 
    events=0x5555557cd900, maxevents=2, timeout=0) at src/common.c:1787
#6  0x00007ffff796d1f1 in sock_pe_progress_rx_ep (pe=0x7ffff7fa6010, 
    ep_attr=0x5555557adfb0, rx_ctx=0x5555557b14f0)
    at prov/sockets/src/sock_progress.c:2392
#7  0x00007ffff796d6fb in sock_pe_progress_rx_ctrl_ctx (pe=0x7ffff7fa6010, 
    rx_ctx=0x5555557b14f0, tx_ctx=0x5555557b12e0)
    at prov/sockets/src/sock_progress.c:2510
#8  0x00007ffff796d8db in sock_pe_progress_tx_ctx (pe=0x7ffff7fa6010, 
    tx_ctx=0x5555557b12e0) at prov/sockets/src/sock_progress.c:2550
#9  0x00007ffff796decf in sock_pe_progress_thread (data=0x7ffff7fa6010)
    at prov/sockets/src/sock_progress.c:2655
#10 0x00007ffff76a36db in start_thread (arg=0x7ffff6a8b700) at pthread_create.c:463
---Type <return> to continue, or q <return> to quit---
```

shefty commented:
Assert is hit because sockets provider has removed the fd from the epoll fd from one thread, while another thread is busy polling for events.  To fix, the assert needs to be converted to an if check, with the event is ignored if the fd is no longer part of the epoll.  There's a higher-level race here that the socket provider must handle, which is handling an event for the fd if the event is reported first.  Socket has some code to do this.

Converted most of the asserts I added before to if statements that simply return if the fd or epoll context is not found.
